### ISSUE TITLE
Improve `FileSystemTransport` by using `compute()` or background isolates

### ISSUE THREAD

**Issue by ueman** [getsentry/sentry-dart#536]
We can try to use [`compute()`](https://api.flutter.dev/flutter/foundation/compute-constant.html) here

https://github.com/getsentry/sentry-dart/blob/bca59abc7f5f257b4fcbb973090d9df719509ebc/flutter/lib/src/file_system_transport.dart#L14-L17

in order to offload the conversion to another isolate.
This could improve the performance because we're not blocking the UI Isolate, especially if there are any attachments or big events.

There are [some things](https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html) which need to be considered.

```dart
class NewFileSystemTransport implements Transport {
  NewFileSystemTransport(this._channel, this._options);

  final MethodChannel _channel;
  final SentryOptions _options;

  @override
  Future<SentryId?> send(SentryEnvelope envelope) async {
    final args = await compute(_convert, envelope);
    try {
      await _channel.invokeMethod<void>('captureEnvelope', [args]);
    } catch (exception, stackTrace) {
      _options.logger(
        SentryLevel.error,
        'Failed to save envelope',
        exception: exception,
        stackTrace: stackTrace,
      );
      return SentryId.empty();
    }

    return envelope.header.eventId;
  }

  static Future<Uint8List> _convert(SentryEnvelope envelope) async {
    final envelopeData = <int>[];
    await envelope.envelopeStream().forEach(envelopeData.addAll);
    // https://flutter.dev/docs/development/platform-integration/platform-channels#codec
    return Uint8List.fromList(envelopeData);
  }
}
```

**Comment by marandaneto** [comment]
makes sense, since Flutter Apps don't crash if not on the Native layer, we could offload the heavy work on a different Isolate using `compute`, worth saying that every capture method returns a `Future`, so they can always choose to not `await` the method anyway.

**Comment by ueman** [comment]
Even if it's not awaited it's still running on the UI isolate.

**Comment by bruno-garcia** [comment]
if we're not awaiting, we need another back pressure mechanism. Otherwise we might degrade performance if the app is producing more events than the SDK can flush out (to disk or network)

**Comment by marandaneto** [comment]
Background isolates might help solving this issue, if we are able to offload to a background Isolate and this Isolate is able to call method channels, only possible on Flutter 3.7 still.

Edit:
3.7 is GA
See https://medium.com/flutter/whats-new-in-flutter-3-7-38cbea71133c

**Comment by denrase** [comment]
Checked the [documentation](https://docs.flutter.dev/platform-integration/platform-channels#using-plugins-and-channels-from-background-isolates) for background channel usage, and we have the same Issue that we are facing with #1801, as we need to access [ServicesBinding.rootIsolateToken](https://api.flutter.dev/flutter/services/ServicesBinding/rootIsolateToken.html), which is static, so we can't use the approach where we cast to dynamic and call a method that is unknown in older Flutter versions which we need to support. Is there approach or a way around this limitation?

**Comment by denrase** [comment]
Short summary from the #1811 findings:

- We can't use compute to convert the envelope, as we capture objects that cannot be shared between isolates. Same issues as in #707
- We'd need to bump flutter to 3.7 in order to use method channels in a background isolate. Is passing the serialised structures to the Cocoa/Android SDKs on the main thread a bottleneck that is causing us issues?

**Comment by buenaflor** [comment]
> Is passing the serialised structures to the Cocoa/Android SDKs on the main thread a bottleneck that is causing us issues?

Have you checked that? 

Not sure if passing serialized data will cause bottleneck issues so I'd think this is just a nice to have in terms of improvement but alone probably doesn't warrant a min version bump to 3.7

**Comment by buenaflor** [comment]
Blocked by min version Flutter 3.7

**Comment by denrase** [comment]
Was trying different approaches today, this is what i found:

- Compute for encoding does not work, as we cannot pass the `SentryEnvelope` to the background isolate.
- We are not using channels anymore to send envelope data to the native side (Android/iOS), so this is already optimal, as the native SDKs will queue envelopes on the background.
- Tried encoding of envelope by moving json encoding to an isolate worker, similar to what we do for replay Android. Implemented various integration tests with large envelopes. Overall duration was slower than current approach, attributing this to the overhead of sending sending every header/item over the ports individually.

<details>
<summary> Utf8JsonEncoderIsolateWorker Code</summary>

```dart
/// A background isolate worker for handling JSON encoding operations.
/// This prevents blocking the main isolate when encoding large JSON objects.
class Utf8JsonEncoderIsolateWorker {
  final SendPort _commands;
  final ReceivePort _responses;
  final Map<int, Completer<dynamic>> _activeRequests = {};
  int _idCounter = 0;
  bool _closed = false;


  static Future<Utf8JsonEncoderIsolateWorker> spawn() async {
    // Create a receive port and add its initial message handler
    final initPort = RawReceivePort();
    final connection = Completer<(ReceivePort, SendPort)>.sync();
    initPort.handler = (SendPort commandPort) {
      connection.complete((
        ReceivePort.fromRawReceivePort(initPort),
        commandPort,
      ));
    };

    // Spawn the isolate.
    try {
      await Isolate.spawn(
        _startRemoteIsolate,
        initPort.sendPort,
        debugName: 'Utf8JsonEncoderWorker',
      );
    } on Object {
      initPort.close();
      rethrow;
    }

    final (ReceivePort receivePort, SendPort sendPort) =
        await connection.future;

    return Utf8JsonEncoderIsolateWorker._(receivePort, sendPort);
  }

  static Object? _jsonSerializationFallback(Object? nonEncodable) {
    if (nonEncodable == null) {
      return null;
    }
    return nonEncodable.toString();
  }

  Utf8JsonEncoderIsolateWorker._(this._responses, this._commands) {
    _responses.listen(_handleResponsesFromIsolate);
  }

  Future<List<int>> convert(Object? object) async {
    if (_closed) {
      throw StateError('Utf8JsonEncoderIsolateWorker is closed');
    }

    final id = _idCounter++;
    final completer = Completer<List<int>>();
    _activeRequests[id] = completer;

    _commands.send((id, object));

    return completer.future;
  }

  /// Batch convert multiple objects in a single isolate call for better performance
  Future<List<List<int>>> convertBatch(List<Object?> objects) async {
    if (_closed) {
      throw StateError('Utf8JsonEncoderIsolateWorker is closed');
    }

    final id = _idCounter++;
    final completer = Completer<List<List<int>>>();
    _activeRequests[id] = completer;

    _commands.send((id, 'batch', objects));

    return completer.future;
  }

  

  void _handleResponsesFromIsolate(dynamic message) {
    final (int id, Object? response) = message as (int, Object?);
    final completer = _activeRequests.remove(id);
    if (completer == null) return;

    if (response is RemoteError) {
      completer.completeError(response);
    } else if (response is List<List<int>>) {
      // Batch response
      completer.complete(response);
    } else {
      // Single response
      completer.complete(response as List<int>);
    }

    if (_closed && _activeRequests.isEmpty) _responses.close();
  }

  void close() {
    if (!_closed) {
      _closed = true;
      _commands.send('shutdown');
      if (_activeRequests.isEmpty) _responses.close();
    }
  }

  /// This is the actual implementation that runs in the background isolate.
  static void _handleCommandsToIsolate(
    ReceivePort receivePort,
    SendPort sendPort,
  ) {
    final utf8JsonEncoder = JsonUtf8Encoder(null, _jsonSerializationFallback, null);
    
    receivePort.listen((message) {
      if (message == 'shutdown') {
        receivePort.close();
        return;
      }
      
      if (message is (int, String, List<Object?>)) {
        // Batch operation
        final (id, command, objects) = message;
        try {
          final results = objects.map((object) => utf8JsonEncoder.convert(object)).toList();
          sendPort.send((id, results));
        } catch (e, stacktrace) {
          sendPort.send((id, RemoteError(e.toString(), stacktrace.toString())));
        }
      } else {
        // Single operation
        final (id, object) = message as (int, Object?);
        try {
          // Encode the object to UTF-8 JSON bytes
          final result = utf8JsonEncoder.convert(object);
          sendPort.send((id, result));
        } catch (e, stacktrace) {
          // Send error back to main isolate - it will handle logging
          sendPort.send((id, RemoteError(e.toString(), stacktrace.toString())));
        }
      }
    });
  }

  static void _startRemoteIsolate(SendPort initPort) {
    final receivePort = ReceivePort();
    initPort.send(receivePort.sendPort);
    _handleCommandsToIsolate(receivePort, initPort);
  }
}
```
</details>

Closing this for now. Please reopen if there's a different suggestions on how to tackle this.
